lib/jose-jws-sign.js

```diff
@@ -0,0 +1,235 @@
+  that.signer_aads = {};
+  that.signer_headers = {};
+};
+
+/**
+ * Adds a signer to JoseJWS instance. It'll be the on of the signers of the resulting JWS.
+ *
+ * @param rsa_key        private RSA key in json format, Parameters can be base64
+ *                       encoded, strings or number (for 'e'). Or CryptoKey
+ * @param key_id         a string identifying the rsa_key. OPTIONAL
+ * @param aad            Object protected header
+ * @param header         Object unprotected header
+ */
+JoseJWS.Signer.prototype.addSigner = function (rsa_key, key_id, aad, header) {
+  var that = this,
+    key_promise = isCryptoKey(rsa_key) ? new Promise(function (resolve) {
+      resolve(rsa_key);
+    }) : Jose.Utils.importRsaPrivateKey(rsa_key, aad && aad.alg ? aad.alg : that.cryptographer.getContentSignAlgorithm(), "sign"),
+    kid_promise;
+
+  if (key_id) {
+    kid_promise = new Promise(function (resolve) {
+      resolve(key_id);
+    });
+  } else if (isCryptoKey(rsa_key)) {
+    throw new Error("key_id is a mandatory argument when the key is a CryptoKey");
+  } else {
+    kid_promise = Jose.WebCryptographer.keyId(rsa_key);
+  }
+
+  that.waiting_kid++;
+
+  return kid_promise.then(function (kid) {
+    that.key_promises[kid] = key_promise;
+    that.waiting_kid--;
+    if (aad) that.signer_aads[kid] = aad;
+    if (header) that.signer_headers[kid] = header;
+    return kid;
+  });
+};
+
+/**
+ * Adds a signature to a JWS object
+ * @param jws JWS Object to be signed or its representation
+ * @param aad     Object protected header
+ * @param header  Object unprotected header
+ * @return Promise<String>
+ */
+JoseJWS.Signer.prototype.addSignature = function (jws, aad, header) {
+
+  var that = this;
+
+  if (Utils.isString(jws)) {
+    jws = JSON.parse(jws);
+  }
+
+  if (jws.payload && Utils.isString(jws.payload) &&
+    jws.protected && Utils.isString(jws.protected) &&
+    jws.header && jws.header instanceof Object &&
+    jws.signature && Utils.isString(jws.signature)) {
+    return that.sign(JWS.fromObject(jws), aad, header);
+  } else {
+    throw new Error("JWS is not a valid JWS object");
+  }
+};
+
+/**
+ * Computes signature.
+ *
+ * @param payload JWS Object or String to be signed
+ * @param aad     Object protected header
+ * @param header  Object unprotected header
+ * @return Promise<JWS>
+ */
+JoseJWS.Signer.prototype.sign = function (payload, aad, header) {
+
+  var that = this,
+    check = Object.keys(that.key_promises).length > 0,
+    kids = [];
+
+  if (!check) {
+    throw new Error("No signers defined. At least one is required to sign the JWS.");
+  }
+
+  if (that.waiting_kid) {
+    throw new Error("still generating key IDs");
+  }
+
+  function sign (message, protectedHeader, unprotectedHeader, rsa_key_promise, kid) {
+
+    var toBeSigned;
+
+    if (!protectedHeader) {
+      protectedHeader = {};
+    }
+
+    if (!protectedHeader.alg) {
+      protectedHeader.alg = cryptographer.getContentSignAlgorithm();
```

Comment:
I know what you mean...
Il 09/Set/2015 13:05, "Alok Menghrajani" <notifications@github.com> ha
scritto:

> In lib/jose-jws-sign.js
> <https://github.com/square/js-jose/pull/15#discussion_r39029895>:
>
> > +  }
> > +
> > +  if (that.waiting_kid) {
> > +    throw new Error("still generating key IDs");
> > +  }
> > +
> > +  function sign (message, protectedHeader, unprotectedHeader, rsa_key_promise, kid) {
> > +
> > +    var toBeSigned;
> > +
> > +    if (!protectedHeader) {
> > +      protectedHeader = {};
> > +    }
> > +
> > +    if (!protectedHeader.alg) {
> > +      protectedHeader.alg = cryptographer.getContentSignAlgorithm();
>
> No, it's fixed in master. Still trying to figure out what's going on with
> travis and jose-jws-test.html, it's driving me nuts :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/js-jose/pull/15/files#r39029895>.
>